TLSv1.3 post-handshake authentication

tmshort opened the pull request:
No longer WIP: from draft-22.
~Need more unit tests (tested via s_client/s_server).~
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [X] documentation is added or updated
- [X] tests are added or updated

tmshort commented:
Still needs tests specific to this functionality.
I'm not in love with the function names as presented here.
I'm also not in love with the "force" API.

FdaSilvaYY commented on ssl/statem/extensions_srvr.c:
```diff
@@ -871,6 +871,17 @@ int tls_parse_ctos_psk(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
     SSL_SESSION_free(sess);
     return 0;
 }
```
new line between the two function bodies

FdaSilvaYY commented on ssl/statem/statem_clnt.c:
```diff
@@ -2431,6 +2451,7 @@ MSG_PROCESS_RETURN tls_process_certificate_request(SSL *s, PACKET *pkt)
                      SSL_R_BAD_LENGTH);
             return MSG_PROCESS_ERROR;
         }
+
```
useful ?

FdaSilvaYY commented on ssl/statem/statem_srvr.c:
```diff
@@ -106,6 +106,14 @@ static int ossl_statem_server13_read_transition(SSL *s, int mt)
         if (s->early_data_state == SSL_EARLY_DATA_READING)
             break;
+
+        if (mt == SSL3_MT_CERTIFICATE &&
```
`&&` goes to the begin of the line

tmshort commented:
> Does this extension deserve some change in t1_trace.c for debug ?

It probably should be put into the `ssl_exts_tbl[]`

tmshort commented:
Unit tests have been added (TLSv1.3 tests in 04-client_auth under 80-test_ssl_new).
More updated docs
Added a CONF parameter.

tmshort commented:
No longer WIP

mattcaswell commented on test/ssl-tests/04-client_auth.conf.in:
```diff
@@ -26,6 +26,7 @@ sub generate_tests() {
         my $caalert;
         my $method;
         my $sctpenabled = 0;
+	my $postauthenabled = 0;
```
Indentation wrong

tmshort replied:
TAB vs SPC, will fix.

mattcaswell commented on test/ssl-tests/04-client_auth.conf.in:
```diff
@@ -39,153 +40,182 @@ sub generate_tests() {
+		    push @tests, {
+			name => "client-auth-${protocol_name}-request"
+                                .($sctp ? "-sctp" : "")
+				.($post ? "-post-handshake" : ""),
```
Indentation. Similar lines below.

tmshort replied:
Will fix.

mattcaswell commented on test/ssl-tests/04-client_auth.conf.in:
```diff
@@ -39,153 +40,182 @@ sub generate_tests() {
+                        },
+                    };
+                    $tests[-1]{"test"}{"UseSCTP"} = "Yes" if $sctp;
+		    if ($post) {
```
Indentation

tmshort replied:
Will fix.

mattcaswell commented on test/ssl-tests/04-client_auth.conf.in:
```diff
@@ -39,153 +40,182 @@ sub generate_tests() {
+                    };
+                    $tests[-1]{"test"}{"UseSCTP"} = "Yes" if $sctp;
+		    if ($post) {
+			$tests[-1]{"server"}{"VerifyMode"} .= "PostHandshake";
```
Is `.=` correct here or should it be just `=`. Same question for similar lines below

tmshort replied:
Yes, when doing post-handshake auth (as indicated by `$post`) the VerifyMode needs to be updated. In this case the `Require` and `Request` method is retained, but appended with `PostHandshake` to end up with `RequirePostHandshake` and `RequestPostHandshake` which sets the appropriate flags via SSL_CONF.

mattcaswell commented on test/ssl-tests/04-client_auth.conf.in:
```diff
@@ -39,153 +40,182 @@ sub generate_tests() {
+                    };
+                    $tests[-1]{"test"}{"UseSCTP"} = "Yes" if $sctp;
+		    $tests[-1]{"server"}{"VerifyMode"} .= "PostHandshake" if $post;
+		    $tests[-1]{"test"}{"HandshakeMode"} = "PostHandshakeAuth" if $post;
```
Indentation

tmshort replied:
Will fix.

tmshort commented:
> It seems that when a client responds to a post-handshake cert request the server sends back a NewSessionTicket afterwards which is unexpected.

That's something I noticed too; I suspect it's because the session state has changed as there is now client authentication information, so a new session ticket is sent. So, while it's unexpected, it's not unreasonable.

> Also, from s_server when requesting a post-handshake cert twice in a row, the second one always fails.

Looks as though the state needs to change once the certificate is received. The intended limitation is that only one request can be outstanding.

tmshort commented on apps/s_server.c:
```diff
@@ -2501,6 +2501,23 @@ static int sv_body(int s, int stype, int prot, unsigned char *context)
+                    continue;
+                    /*
+                     * strcpy(buf,"server side POST-HANDSHAKE asking for client
+                     * cert\n");
```
I was being consistent! :)

tmshort commented on include/openssl/ssl.h:
```diff
@@ -1056,6 +1056,7 @@ size_t SSL_get_peer_finished(const SSL *s, void *buf, size_t count);
 # define SSL_VERIFY_PEER                 0x01
 # define SSL_VERIFY_FAIL_IF_NO_PEER_CERT 0x02
 # define SSL_VERIFY_CLIENT_ONCE          0x04
+# define SSL_VERIFY_POST_HANDSHAKE       0x08
```
Will fix.

tmshort commented on ssl/ssl_locl.h:
```diff
@@ -1283,6 +1291,7 @@ struct ssl_st {
          * as this extension is optional on server side.
          */
         uint8_t max_fragment_len_mode;
+
```
Will fix.

tmshort commented on ssl/ssl_locl.h:
```diff
@@ -2520,6 +2535,10 @@ __owur int srp_generate_server_master_secret(SSL *s);
 
+/* statem/statem_srvr.c */
+
+int send_certificate_request(SSL *s);
```
Will fix.

tmshort commented on ssl/statem/extensions.c:
```diff
@@ -1556,3 +1568,18 @@ static int final_maxfragmentlen(SSL *s, unsigned int context, int sent)
+static int final_post_handshake_auth(SSL *s, unsigned int context, int sent)
+{
+    if (s->server && sent)
+        s->post_handshake_auth = SSL_PHA_EXT_RECEIVED;
```
Given the condition under which the variable is set, yes.

mattcaswell commented:
> That's something I noticed too; I suspect it's because the session state has changed as there is now client authentication information, so a new session ticket is sent. So, while it's unexpected, it's not unreasonable.

The reason is that it happens automatically after the server receives a client Finished. Since we've received another one, it sends another ticket. Like you say though, perhaps this is ok?? If we keep that behaviour perhaps it should be documented somewhere (not sure where).

tmshort commented on ssl/statem/extensions_clnt.c:
```diff
@@ -1133,6 +1133,44 @@ EXT_RETURN tls_construct_ctos_psk(SSL *s, WPACKET *pkt, unsigned int context,
+
+        for (i = 0; i < SSL_PKEY_NUM; i++)
+            if (s->cert->pkeys[i].x509 != NULL && s->cert->pkeys[i].privatekey != NULL)
+            n++;
```
Will fix.

mattcaswell commented:
Another thought: how does a server application know when a certificate has arrived?

tmshort commented on ssl/statem/statem_srvr.c:
```diff
@@ -324,11 +332,17 @@ static int send_server_key_exchange(SSL *s)
+           && (!SSL_IS_TLS13(s) || !(s->verify_mode & SSL_VERIFY_POST_HANDSHAKE)
+               || s->post_handshake_auth == SSL_PHA_REQUEST_PENDING)
            /*
             * if SSL_VERIFY_CLIENT_ONCE is set, don't request cert
```
`SSL_VERIFY_CLIENT_ONCE`, effectively disables post-handshake authentication; since it prevents a certificate request from happening during a renegotiation (anything after the initial handshake). To make it work with post-handshake authentication, I'm thinking a certificate received flag/counter might be needed.

tmshort replied:
So, I ever-so-slightly changed the behavior of SSL_VERIFY_CLIENT_ONCE to only send a CertReq once (period). This behavior is mostly consistent with the existing behavior; especially since the s_server.c code sets the SSL_VERIFY_CLIENT_ONCE flag during cert renegotiation (when it receives `GET /renegcert` which shouldn't work according to the documentation). That means that post-handshake auth will fail immediately if certificate authentication previously occurred during the initial handshake, or will fail the second time if certificate authentication did not occur during the initial handshake.

tmshort commented on ssl/statem/statem_srvr.c:
```diff
@@ -422,7 +440,11 @@ static WRITE_TRAN ossl_statem_server13_write_transition(SSL *s)
+        if (s->post_handshake_auth == SSL_PHA_REQUEST_PENDING) {
+            s->post_handshake_auth = SSL_PHA_REQUESTED;
+            st->hand_state = TLS_ST_OK;
+        } else
```
Will fix.

tmshort commented on ssl/statem/statem_srvr.c:
```diff
@@ -2686,12 +2715,25 @@ int tls_construct_server_key_exchange(SSL *s, WPACKET *pkt)
+            OPENSSL_free(s->pha_context);
+            s->pha_context_len = 32;
+            if ((s->pha_context = OPENSSL_malloc(s->pha_context_len)) == NULL ||
+                    RAND_bytes(s->pha_context, s->pha_context_len) <= 0 ||
```
New to me... will fix.

mattcaswell commented:
Just noticed I get "Message length parse error!" when using -trace to s_server/s_client and a CertificateRequest is sent. Not sure if it is specific to this PR or whether it is a pre-existing issue.

tmshort commented on ssl/statem/statem_clnt.c:
```diff
@@ -131,6 +131,10 @@ static int ossl_statem_client13_read_transition(SSL *s, int mt)
         break;
 
     case TLS_ST_CR_CERT_REQ:
+        if (s->post_handshake_auth == SSL_PHA_REQUESTED) {
```
Yes, I see that. After receiving a CertRequest in post-handshake auth, there should be no further messages from the server, and the client should just send it's Certificate/CertVerify/Finished sequence. But, I'm not seeing it being hit in my test, so it appears that it can be removed.

mattcaswell replied:
Yes, it shouldn't ever get here. This point in the code is about working out what state to move to if we read another message after the cert req. But after a cert req message is received it goes immediately into the writing state so we never attempt to read another message. This looks like dead code.

tmshort replied:
Agreed.

mattcaswell commented:
Some thoughts on Finished construction:

- cleanuphand is set which has the effect of incrementing sess_hit counters and other tasks which are supposed to only happen at the end of a "proper" handshake
- It's not clear to me that hash calculation is correct. Which messages are actually being included in the transcript hash in this scenario? The spec says "ClientHello ... client  Finished + CertificateRequest". I'm wondering what happens if there have been NewSessionTickets, or KeyUpdates etc in the meantime?
- It looks like in the Finished post work we do stuff like change the cipher state, which will have the effect of incorrectly resetting the sequence counters etc.

Have you done any interoperability testing with this PR? I suspect some of the above issues would fail that kind of testing.

mattcaswell commented:
>  The spec says "ClientHello ... client Finished + CertificateRequest".

Oh - plus of course Certificate and CertificateVerify

tmshort commented:
> Some thoughts on Finished construction:
>
> * cleanuphand is set which has the effect of incrementing sess_hit counters and other tasks which are supposed to only happen at the end of a "proper" handshake

I'll take a look at not futzing with this state (as with the comment below).

> * It's not clear to me that hash calculation is correct. Which messages are actually being included in the transcript hash in this scenario? The spec says "ClientHello ... client Finished + CertificateRequest". I'm wondering what happens if there have been NewSessionTickets, or KeyUpdates etc in the meantime?

Yes, not entirely clear. Are NewSessionTickets and KeyUpdates added to the transcript hash?

> * It looks like in the Finished post work we do stuff like change the cipher state, which will have the effect of incorrectly resetting the sequence counters etc.
>
> Have you done any interoperability testing with this PR? I suspect some of the above issues would fail that kind of testing.

I have not done interoperability testing; do we know of any other implementations? @davidben ?

mattcaswell commented:
> Yes, not entirely clear. Are NewSessionTickets and KeyUpdates added to the transcript hash?

All messages we read are fed in automatically as part of tls_get_message_body() with a special exception for HRRs because of the synthetic message that gets added - so yes I think they are.

However even if we put in an exception for NewSessionTicket and KeyUpdate there is still an issue if there are multiple post-handshake CertificateRequests. See the spec:

    Note, however, that subsequent post-handshake authentications do not include each other,
    just the messages through the end of the main handshake.

So we should always reset the transcript hash back to its state at the end of the initial handshake before processing a new post-handshake auth sequence (either at the end of the previous sequence, or at the start of the new one).

tmshort commented:
> So we should always reset the transcript hash back to its state at the end of the initial handshake before processing a new post-handshake auth sequence (either at the end of the previous sequence, or at the start of the new one).

Makes sense.

mattcaswell commented:
Another thought: in `tls_process_new_session_ticket()` we have this comment:
````
    /*
     * Sessions must be immutable once they go into the session cache. Otherwise
     * we can get multi-thread problems. Therefore we don't "update" sessions,
     * we replace them with a duplicate. In TLSv1.3 we need to do this every
     * time a NewSessionTicket arrives because those messages arrive
     * post-handshake and the session may have already gone into the session
     * cache.
     */
````

However in `tls_process_client_certificate()` we see this code:

````
    X509_free(s->session->peer);
    s->session->peer = sk_X509_shift(sk);
    s->session->verify_result = s->verify_result;

    sk_X509_pop_free(s->session->peer_chain, X509_free);
    s->session->peer_chain = sk;
````

So there is a potential multi-thread issue with this code now that `tls_process_client_certificate()` is called post-handshake after the session is already in the cache.

tmshort commented:
> So there is a potential multi-thread issue with this code now that tls_process_client_certificate() is called post-handshake after the session is already in the cache.

`tls_process_new_session_ticket() is handled on the client, `tls_process_client_certificate()` is handled on the server: two different session caches. Same rules apply? Should this happen all the time? Renegotiation w/ certificate would also cause a problem here.

mattcaswell commented:
> tls_process_new_session_ticket() is handled on the client, tls_process_client_certificate() is handled on the server: two different session caches. Same rules apply? Should this happen all the time? Renegotiation w/ certificate would also cause a problem here.

Well, I think the point is that once a session is available for use by other threads then the problem could occur. Consider a server application which receives two requests simultaneously using the same session. Both would use the same session from the cache and therefore problems could occur I think?

tmshort commented:
It was really more of a question of having to handle this generically, rather than just for post-handshake auth. Which I think is the case.

mattcaswell commented:
> Renegotiation w/ certificate would also cause a problem here.

Actually, I think not because it would only be an issue if we were resuming (and hence using the same session) - but reneg with certificate request implies full handshake.

davidben commented:
> I have not done interoperability testing; do we know of any other implementations? @davidben ?

We have no plans to implement this feature.

tmshort commented:
>> I have not done interoperability testing; do we know of any other implementations? davidben ?

> We have no plans to implement this feature.

That rules out BoringSSL and Go?
PolarSSL, LibreSSL, and GnuTLS do not seem to have any traction on draft-22 (although GnuTLS might be the closest with draft-21 support). Can't quite tell where NSS is; they have a draft-19 branch, but no mention of draft-22.

tmshort commented:
> * It looks like in the Finished post work we do stuff like change the cipher state, which will have the effect of incorrectly resetting the sequence counters etc.

As the code is now; it appears both sides are making the same change, and everything "just works" (i.e. both sides can send/receive). If I stop the server-side from making these changes (in `tls_process_finished()`, the server can no longer receive data (but can send). So yeah, counters, etc., seem to be reset.

richsalz commented:
@davidben which we?  Will Chrome handle post-auth cert requests?  Even if boring-using servers will never generate them?

davidben commented:
Neither.

davidben commented:
(More accurately, Chrome does not advertise the extension to begin with, so it is forbidden for the server to request it at all.)

tmshort commented:
@mattcaswell The only thing left, I believe, is updating the transcript hash handling? I'll be looking into that later.

mattcaswell commented:
I haven't seen a response to this question yet:

> Another thought: how does a server application know when a certificate has arrived?

tmshort commented:
> Another thought: how does a server application know when a certificate has arrived?

verify callback.

mattcaswell commented:
> I'm seeing the "Message length parse error!" on the master branch; so unrelated to this change.

Fix in #5008.

tmshort commented:
Not sure why the travis tests are failing. At first glance it seems unrelated, but not sure.

mattcaswell commented:
Looks like something fairly fundamental in travis failed:

````
$ clang-3.9 --version
clang-3.9: command not found
````

i.e. Travis failure looks unrelated to this PR

tmshort commented on doc/man3/SSL_CTX_set_verify.pod:
```diff
@@ -43,6 +48,15 @@ verification that shall be allowed for B<ctx>.
+
+SSL_verify_client_post_handshake() causes a Certificate Request message to be sent
+by a server on the given B<ssl> connection. The SSL_VERIFY_PEER flag must be set,
+the SSL_VERIFY_POST_HANDSHAKE flag is optional.
```
The `SSL_VERIFY_POST_HANDSHAKE` flag blocks certificate verification in the initial handshake. It can be used to configure an SSL_CTX or SSL before establishing the connection, rather than setting the flags later when calling `SSL_verify_client_post_handshake()`. It's also a convenience for unit tests; removing it would complicate the unit tests.

mattcaswell replied:
Ah, ok. I understand it now. The docs do say that, but somehow I didn't get that understanding from them. Probably that's just me.

mattcaswell replied:
Should the docs say what happens if you try this on a non TLSv1.3 connection?

tmshort commented on ssl/statem/statem_clnt.c:
```diff
@@ -160,6 +160,17 @@ static int ossl_statem_client13_read_transition(SSL *s, int mt)
+        if (mt == SSL3_MT_CERTIFICATE_REQUEST) {
+            if (s->post_handshake_auth == SSL_PHA_EXT_SENT) {
+                s->post_handshake_auth = SSL_PHA_REQUESTED;
+                if (!tls13_restore_handshake_digest_for_pha(s)) {
```
Well, that's interesting. While I could put a `#error` here, would it be better to fix the original problem? Perhaps the handshake digest should be restored just before adding the message? Would it make sense to have a separate digest storage for post-auth in addition to storing the digest state through clientFinished (not sure; I didn't do it that way to avoid complicating the CertRequest/Certificate/CertVerify/Finished and the extra code to extract and generate it)?

tmshort replied:
A thought: if KeyUpdate and NewSessionTicket are no longer added to the cache, might restoring the digest state after the client Finish also work?

mattcaswell replied:
> Well, that's interesting. While I could put a #error here, would it be better to fix the original problem?

Ideally yes that would be nice (although not essential since there will no doubt be lots of other changes required for DTLS1.3)

> Perhaps the handshake digest should be restored just before adding the message?

Possibly. `ssl3_finish_mac()` could be a good place since it is called by both TLS and DTLS

> A thought: if KeyUpdate and NewSessionTicket are no longer added to the cache, might restoring the digest state after the client Finish also work?

Yes that could work. Although I think my preference would be to reset it at the beginning of a new post-handshake auth rather than after client Finish.

tmshort commented on ssl/statem/statem_clnt.c:
```diff
@@ -424,9 +442,12 @@ static WRITE_TRAN ossl_statem_client13_write_transition(SSL *s)
 
+    case TLS_ST_CW_FINISHED:
+        if (s->post_handshake_auth == SSL_PHA_REQUESTED)
+            s->post_handshake_auth = SSL_PHA_EXT_SENT;
```
Agreed; I put it here because it worked, but it didn't seem like the right place. Will move.

tmshort commented on ssl/statem/statem_clnt.c:
```diff
@@ -3354,8 +3384,12 @@ WORK_STATE tls_prepare_client_certificate(SSL *s, WORK_STATE wst)
         }
-        if (ssl3_check_client_certificate(s))
+        if (ssl3_check_client_certificate(s)) {
+            if (s->post_handshake_auth == SSL_PHA_REQUESTED) {
```
I will check.

tmshort replied:
Not so sure. If `ssl3_check_client_certificate()` returns 1 (i.e. there's a good certificate), then it doesn't go into `WORK_MORE_B`, which this PR does not change. I think the real problem is that if `ssl3_check_client_certificate()` returns 0, it will fall into `WORK_MORE_B` state which won't do the transition of `post_handshake_auth` to `SSL_PHA_REQUESTED`

tmshort replied:
If `WORK_FINISHED_STOP` is not returned during PHA, then the client will expect there to be  Certificate message following the CertificateRequest, which won't be happening with PHA.

mattcaswell replied:
Right....so the the `WORK_MORE_B` state code needs to change to return `WORK_FINISHED_STOP` in the event of `ssl3_check_client_certificate()` returning 0 and we're in PHA?

tmshort replied:
Yes, otherwise, the client waits for a Certificate message that never comes, and doesn't respond. I have additional commits with fixes forthcoming.

tmshort commented on ssl/statem/statem_lib.c:
```diff
@@ -43,12 +43,12 @@ int ssl3_do_write(SSL *s, int type)
         /*
          * should not be done for 'Hello Request's, but in that case we'll
          * ignore the result anyway
+         * TODO(TLS1.3): KeyUpdate and NewSessionTicket do not need to be added
```
Which is why I put it as a TODO, rather than adding the code. But your comment below may mean we should do it.

tmshort commented on ssl/statem/statem_srvr.c:
```diff
@@ -2686,12 +2717,30 @@ int tls_construct_server_key_exchange(SSL *s, WPACKET *pkt)
+                return 0;
+            }
+            /* reset the handshake hash back to just after the ClientFinished */
+            if (!tls13_restore_handshake_digest_for_pha(s)) {
```
Yup; not easy to (unit) test this. But, perhaps this means that the KeyUpdate and NewSessionTicket should not be placed into the messaging digest.

mattcaswell replied:
> Yup; not easy to (unit) test this.

It could be done. Create a connection using `create_ssl_connection()`. Call `SSL_verify_client_post_handshake()` followed by `SSL_do_handshake()` on the server side. On the client side call SSL_key_update() and SSL_do_handshake() without first reading anything from the connection. Then attempt to read on the client side so it sees and processes the CertificateRequest message.

> But, perhaps this means that the KeyUpdate and NewSessionTicket should not be placed into the messaging digest.

Or perhaps we always restore the message digest state at the beginning of a post-handshake sequence

tmshort replied:
If the server receives KeyUpdate between sending the CertRequest and receiving the Certificate... messages (as suggested), it's digest state will be incorrect. So, regardless of when the digest state is restored (which I agree, should be at the latest possible moment), I think KeyUpdate and NewSessionTicket should be excluded from the digests.

mattcaswell replied:
Oh...right. The penny just dropped that the CertRequest itself is included in the transcript hash. So yes, you are right.

tmshort replied:
A unit test has been added for this in sslapitest.c

tmshort commented:
> Create a connection using create_ssl_connection(). Call SSL_verify_client_post_handshake() followed by SSL_do_handshake() on the server side. On the client side call SSL_key_update() and SSL_do_handshake() without first reading anything from the connection. Then attempt to read on the client side so it sees and processes the CertificateRequest message.

I'm thinking of creating a new set of `test_ssl_new` tests for PHA, separating them from the existing `04-client_auth` tests.

mattcaswell commented:
> I'm thinking of creating a new set of test_ssl_new tests for PHA, separating them from the existing 04-client_auth tests.

That would be fine (good idea). However, you might find for the corner cases like this one a test more along the lines of `sslapitest` might work better. Your choice though.

tmshort commented:
I think this is basically done.
I haven't rebased this, as there's been a lot of updates. Of course, I can squash into a single commit and rebase on top of the latest master.

tmshort commented:
Rebased due to conflict.

mattcaswell commented on CHANGES:
```diff
@@ -9,6 +9,9 @@
 
  Changes between 1.1.0f and 1.1.1 [xx XXX xxxx]
 
+  *) Add TLSv1.3 post-handshake authentication.
```
As an aside, we haven't actually got a CHANGES entry for adding TLSv1.3 support yet!! I should probably create a PR.

tmshort replied:
I can remove this, if you'd prefer a "bigger" TLSv1.3 addition to `CHANGES`

mattcaswell replied:
No - this was an aside only. I will add a separate TLSv1.3 entry, but there's no reason not to have this one too.

mattcaswell commented on apps/s_server.c:
```diff
@@ -962,6 +964,7 @@ const OPTIONS s_server_options[] = {
     {"max_early_data", OPT_MAX_EARLY, 'n',
      "The maximum number of bytes of early data"},
     {"early_data", OPT_EARLY_DATA, '-', "Attempt to read early data"},
+    {"force_pha", OPT_FORCE_PHA, '-', "Force-enable post-handshake-authentication (violates protocol)"},
```
Why does this violate the protocol?

tmshort replied:
Because it allows the server to send the post-handshake authentication request without receiving the post-handshake extensions; to be used for testing.

mattcaswell commented on apps/s_client.c:
```diff
@@ -787,6 +788,7 @@ const OPTIONS s_client_options[] = {
 #endif
     {"keylogfile", OPT_KEYLOG_FILE, '>', "Write TLS secrets to file"},
     {"early_data", OPT_EARLY_DATA, '<', "File to send as early data"},
+    {"force_pha", OPT_FORCE_PHA, '-', "Force-enable post-handshake-authentication"},
```
This option (and the s_server equivalent) needs adding to the docs

tmshort replied:
Added to docs, s_server option removed.

mattcaswell commented on apps/s_server.c:
```diff
@@ -2208,6 +2214,9 @@ static int sv_body(int s, int stype, int prot, unsigned char *context)
     }
 
+    if (s_force_pha)
+        SSL_force_post_handshake_auth(con);
```
What does this actually do on the server side? The docs seem to indicate its a client side option.

tmshort replied:
I added it to the server-side to permit testing; allows the server to initiate a post-handshake auth without receiving the extension.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -5291,3 +5293,58 @@ __owur unsigned int ssl_get_split_send_fragment(const SSL *ssl)
+    }
+
+    if (ssl->pha_forced)
+        ssl->post_handshake_auth = SSL_PHA_EXT_RECEIVED;
```
Ahhh...now I understand what `SSL_force_post_handshake_auth()` does on the server side. Why do we need this? Is it wise?

tmshort replied:
Testing.

mattcaswell commented on ssl/statem/statem_clnt.c:
```diff
@@ -160,6 +160,26 @@ static int ossl_statem_client13_read_transition(SSL *s, int mt)
         }
+        if (mt == SSL3_MT_CERTIFICATE_REQUEST) {
+#if DTLS_MAX_VERSION != DTLS1_2_VERSION
+#error TODO(DTLS1.3): Restore digest for PHA before adding message.
```
Nit: should be `# error`, i.e. with a space

mattcaswell commented on ssl/statem/statem_srvr.c:
```diff
@@ -3385,16 +3435,24 @@ MSG_PROCESS_RETURN tls_process_client_certificate(SSL *s, PACKET *pkt)
+    }
+
+    if (!PACKET_get_net_3(pkt, &llen)
             || !PACKET_get_sub_packet(pkt, &spkt, llen)
```
Probably we should be using `PACKET_as_length_prefixed_3` here I think?

tmshort replied:
Maybe? The `PACKET_get_net_3()` was part of the original code.

mattcaswell replied:
Yeah, I know - but since your touching it anyway...

tmshort commented:
Github won't let me respond to this comment??
> Should the docs say what happens if you try this on a non TLSv1.3 connection?
It should say TLSv1.3 only.

tmshort commented:
I can remove the s_server option, or leave it for testing. However, I think the API should remain.

mattcaswell commented:
> I can remove the s_server option, or leave it for testing. However, I think the API should remain.

I'm not a fan of adding API which allows you to break the protocol specs, even if that is for testing purposes. Its kind of a dangerous precedent to set IMO.

richsalz commented:
I agree with Matt, remove the API or at least "#if 0" it.

tmshort commented:
> I agree with Matt, remove the API or at least "#if 0" it.

The API still applies to the client; I will need to figure out a way to allow it to work for unit-testing purposes.

tmshort commented:
Rebased (to get around the C90 issue) and added another commit for the latest nits.

tmshort commented:
Having some off travis failures; one is a failed `apt-get`, which seems unrelated. The other appears to be related to doc-nits, but I'm not seeing that, so, not sure what's going on there?

tmshort commented:
Rebased and squashed, as changes were made and reverted.

kaduk commented on test/ssl-tests/26-tls13_client_auth.conf.in:
```diff
@@ -0,0 +1,293 @@
+        },
+        test => {
+            "ExpectedResult" => "ServerFail",
+            "HandshakeMode" => "PostHandshakeAuth",
```
I guess I'm unclear on why this is supposed to fail.  The client has no cert configured, sure, but the server is only requesting post-handshake, not requiring it.

tmshort replied:
The client has no cert configured, so it does not sent the extension; the server's attempt to do a post-handshake auth will not occur, and thus, fail.

kaduk replied:
Ah.  And there's no expected alert to check for like in the other cases?

tmshort replied:
No, the call to SSL_verify_client_post_handshake() just fails.

kaduk commented on test/sslapitest.c:
```diff
@@ -3217,6 +3217,65 @@ static int test_max_fragment_len_ext(int idx_tst)
+        goto end;
+
+    /* Starts with SSL_connect(), but it's really just SSL_do_handshake() */
+    if (!TEST_true(create_ssl_connection(serverssl, clientssl,
```
I guess I would consider renaming create_ssl_connection() to indicate its more generic usage, but don't have any good ideas for alternate names.

tmshort replied:
create_ssl_connection() is pre-existing test code; out of scope.

kaduk replied:
I think this might be the first time that create_ssl_connection() is (ab)used for driving the state machine after the initial handshake; if so, this represents and expansion of its "documented scope" (not that it's documented) and could justify renaming the function to reflect its new broader scope.

tmshort replied:
I still think it's out of scope.

tmshort commented on doc/man3/SSL_CTX_set_verify.pod:
```diff
@@ -43,6 +48,16 @@ verification that shall be allowed for B<ctx>.
 
+SSL_force_post_handshake_auth() forces the Post-Handshake Authentication extension
+to be added to the ClientHello regardless of certificate configuration at the time
+of the initial handshake, such that post-handshake authentication can be requested
```
But they all line up!

tmshort replied:
I will limit them to 80.

tmshort commented on ssl/statem/statem_lib.c:
```diff
@@ -755,8 +758,14 @@ MSG_PROCESS_RETURN tls_process_finished(SSL *s, PACKET *pkt)
+            s->statem.cleanuphand = 1;
+        if (SSL_IS_TLS13(s) && !tls13_save_handshake_digest_for_pha(s)) {
+                /* SSLfatal() already called */
+                return MSG_PROCESS_ERROR;
```
Yes, that's the intent.

tmshort commented on ssl/statem/statem_srvr.c:
```diff
@@ -324,16 +331,22 @@ static int send_server_key_exchange(SSL *s)
+            * post-handshake in TLSv1.3:
+            */
+           && (!SSL_IS_TLS13(s) || !(s->verify_mode & SSL_VERIFY_POST_HANDSHAKE)
+               || s->post_handshake_auth == SSL_PHA_REQUEST_PENDING)
```
Correct, that is the intent of that flag.

tmshort commented on ssl/statem/statem_srvr.c:
```diff
@@ -3500,13 +3557,45 @@ MSG_PROCESS_RETURN tls_process_client_certificate(SSL *s, PACKET *pkt)
     s->session->peer_chain = sk;
 
+    if (new_sess != NULL)
+        ssl_update_cache(s, SSL_SESS_CACHE_SERVER);
```
I'm not sure that matters? This change doesn't impact the behavior.

kaduk replied:
Right, this change doesn't affect the behavior.  But a reader might wonder why only this one case gets inserted into the cache -- "shouldn't the session for non-PHA get cached, too?"

tmshort replied:
Still not sure it matters; IMHO, the session is not complete until the handshake is done, as per the relevant RFCs. I think it's out-of-scope, since it's the the normal handshake.

tmshort commented on ssl/statem/statem_srvr.c:
```diff
@@ -422,7 +439,12 @@ static WRITE_TRAN ossl_statem_server13_write_transition(SSL *s)
     case TLS_ST_SW_CERT_REQ:
-        st->hand_state = TLS_ST_SW_CERT;
+        if (s->post_handshake_auth == SSL_PHA_REQUEST_PENDING) {
+            s->post_handshake_auth = SSL_PHA_REQUESTED;
```
This state (TLS_ST_SW_CERT_REQ) is pre-existing, and is used by the server to write out the certificate request in any part of the handshake.

kaduk replied:
I'm asking about SSL_PHA_REQUEST_PENDING vs. SSL_PHA_REQUESTED.
Both of those indicate that we are in the middle of a PHA flow, and my intuition says that there should only be a codepath that gets here with one of those values set, and the other conditional is dead code.

tmshort replied:
The else case here is the normal (initial) handshake case. The only way to get into this state during the initial handshake (where the value could be SSL_PHA_NONE or SSL_PHA_EXT_RECEIVED), or if we're in SSL_PHA_REQUEST_PENDING.

tmshort replied:
We won't get here if we're SSL_PHA_REQUESTED.

tmshort commented on doc/man3/SSL_CTX_set_verify.pod:
```diff
@@ -154,6 +182,20 @@ Its return value is identical to B<preverify_ok>, so that any verification
+must be called before SSL_connect().
+
+SSL_verify_client_post_handshake() requires that verify flags have been previously
+set, and that a client sent the post-handshake authentication extension.
```
I'll add it to the RETURN VALUES section.

tmshort commented on apps/s_server.c:
```diff
@@ -2496,10 +2489,19 @@ static int sv_body(int s, int stype, int prot, unsigned char *context)
-                    /*
-                     * strcpy(buf,"server side RE-NEGOTIATE asking for client
-                     * cert\n");
-                     */
```
Unless it misunderstood, it was a request from @mattcaswell. Of course, I can just put them back, and someone else can remove them.

mattcaswell replied:
Yes these should go. But I think @kaduk was asking for a separate *commit* not necessary a separate PR.

tmshort replied:
It's now a separate commit.

tmshort commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -435,6 +435,18 @@ occurs if the client does not present a certificate. Servers only.
+not require a certificate from the client post-handshake. A certificate will
+not be requested during the initial handshake. The server application must
+provide a mechanism to request a certificate post-handshake. Servers only.
+TLSv1.3 only.
```
Yes it could; this provide an alternative mechanism via configuration. It's certainly used during the unit tests.

kaduk commented on ssl/statem/statem_lib.c:
```diff
@@ -2191,3 +2211,54 @@ size_t construct_key_exchange_tbs(SSL *s, unsigned char **ptbs,
+                 ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (!EVP_MD_CTX_copy_ex(s->s3->handshake_dgst,
```
Okay.
(I think this could happen if ssl3_init_finished_mac() got called at an inopportune time, but I think that would be a programming error if it happened.)

tmshort commented:
Rebased, added the comment-only commit.

tmshort commented:
> OPT_R_ENUM and the NULL check are the main actionable things, I think.

Fixed the enum issue, and the NULL check was determined to not be necessary.

mattcaswell commented:
Ping @openssl/committers for second review

tmshort commented:
Rebased and squashed. Minor changes to deal with updated master:
* Removed change to CHANGES
* Updated test/build.info
* Updated function numbers in sslerr.h and openssl.txt

tmshort commented on ssl/ssl_lib.c:
```diff
@@ -1186,6 +1186,8 @@ void SSL_free(SSL *s)
     OPENSSL_free(s->ext.tls13_cookie);
     OPENSSL_free(s->clienthello);
+    OPENSSL_free(s->pha_context);
+    EVP_MD_CTX_free(s->pha_dgst);
```
But in the case of `s->pha_dgst`, it might be necessary to free in `SSL_clear()` since the SSL object could be reused for a new handshake. The use of `pha_dgst` assumes that it is allocated exactly once per handshake after the first Finished exchange.
`pha_context` is less critical, as it's reallocated as needed.

tmshort commented on doc/man3/SSL_CTX_set_verify.pod:
```diff
@@ -154,6 +183,20 @@ Its return value is identical to B<preverify_ok>, so that any verification
 
+After calling SSL_force_post_handshake_auth(), the client will need to add a
+certificate to its configuration before it can successfully authenticate. This
+must be called before SSL_connect().
```
Or a certificate callback.

tmshort commented on doc/man3/SSL_CTX_set_verify.pod:
```diff
@@ -154,6 +183,20 @@ Its return value is identical to B<preverify_ok>, so that any verification
+sent to the client, this can be done with SSL_do_handshake() or SSL_write_ex().
+Only one certificate request may be outstanding at any time.
+
+When post-handshake authentication occurs, a refreshed B<NewSessionTicket>
```
probably not

tmshort commented on ssl/statem/statem_clnt.c:
```diff
@@ -3332,6 +3367,7 @@ static int ssl3_check_client_certificate(SSL *s)
     if (s->cert->cert_flags & SSL_CERT_FLAGS_CHECK_TLS_STRICT &&
         !tls1_check_chain(s, NULL, NULL, NULL, -2))
         return 0;
+
```
cruft from code that was added at one point, I guess.

tmshort commented on ssl/statem/statem_clnt.c:
```diff
@@ -375,6 +395,13 @@ static WRITE_TRAN ossl_statem_client13_write_transition(SSL *s)
+
     default:
         /* Shouldn't happen */
         SSLfatal(s, SSL_AD_INTERNAL_ERROR,
```
There are at least 2 other /* Fall through */ conditions in this switch. However, the default case here and other locations is originally at the top, and I'm assuming for a reason. So, I can move this to the bottom and not fall through.

tmshort commented:
Fixed some nits.

kaduk commented:
Thanks for the updates; they look fine to me.
@mattcaswell I think the ball's back in your court

mattcaswell commented:
Pushed. Thanks.
